sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerUIEventListener.scala

```diff
@@ -0,0 +1,204 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive.thriftserver.ui
+
+import org.apache.hive.service.cli.SessionHandle
+import org.apache.hive.service.cli.session.HiveSession
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.scheduler.{SparkListenerJobStart, SparkListener}
+
+import scala.collection.mutable.HashMap
+
+private[thriftserver] trait ThriftServerEventListener {
+  /**
+   * Called when a session created.
+   */
+  def onSessionCreated(ip: String, session: HiveSession) { }
+
+  /**
+   * Called when a session closed.
+   */
+  def onSessionClosed(session: HiveSession) { }
+
+  /**
+   * Called when a statement started to run.
+   */
+  def onStatementStart(id: String, session: HiveSession, statement: String) { }
+  /**
+   * Called when a statement completed compilation.
+   */
+  def onStatementParse(id: String, executePlan: String) { }
+  /**
+   * Called when a statement got a error during running.
+   */
+  def onStatementError(id: String, errorMessage: String, errorTrace: String) { }
+  /**
+   * Called when a statement ran success.
+   */
+  def onStatementFinish(id: String) { }
+}
+
+private[thriftserver] class SessionInfo(
+    val session: HiveSession,
+    val startTimestamp: Long,
+    val ip: String) {
+  val sessionID = session.getSessionHandle.getSessionId.toString
+  val userName = if(session.getUserName == null) "UNKNOWN" else session.getUserName
+  var finishTimestamp = 0L
+  var totalExecute = 0
+
+  def totalTime = {
+    if (finishTimestamp == 0L) {
+      System.currentTimeMillis() - startTimestamp
+    } else {
+      finishTimestamp - startTimestamp
+    }
+  }
+}
+
+private[thriftserver] object ExecutionState extends Enumeration {
+  val STARTED, COMPILED, FAILED, FINISHED = Value
+  type ExecutionState = Value
+}
+
+private[thriftserver] class ExecutionInfo(
+    val statement: String,
+    val session: HiveSession,
+    val startTimestamp: Long) {
+  val userName = if(session.getUserName == null) "UNKNOWN" else session.getUserName
+  var finishTimestamp = 0L
+  var executePlan = ""
+  var detail = ""
+  var state: ExecutionState.Value = ExecutionState.STARTED
+  var jobId = ""
+  var groupId = ""
+  def totalTime = {
+    if (finishTimestamp == 0L) {
+      System.currentTimeMillis() - startTimestamp
+    } else {
+      finishTimestamp - startTimestamp
+    }
+  }
+}
+
+private[sql] class ThriftServerUIEventListener(val conf: SparkConf)
+  extends ThriftServerEventListener with SparkListener {
+
+  import ThriftServerUIEventListener._
+
+  var sessionList = new HashMap[SessionHandle, SessionInfo]
+  var executeList = new HashMap[String, ExecutionInfo]
+  val retainedStatements =
+    conf.getInt("spark.thriftserver.ui.retainedStatements", DEFAULT_RETAINED_STATEMENTS)
```

Comment:
If you're going to use this strategy to limit memory usage, check out the memory leak prevention tests that I wrote for JobProgressListener: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala#L85.  The main idea was to partition the state according to how it grows / shrinks, then write some general assertions.  This may be overkill for what you're doing here, but I just thought that I should mention that approach since it's a nice testing strategy.